### ISSUE TITLE
Don't depend on unstable AndroidX libs

### ISSUE THREAD

**Issue by BenTilbrook** [ReactiveCircus/FlowBinding#91]
`flowbinding-recyclerview` `0.11.1` currently depends on `androidx.recyclerview:1.2.0-alpha02`. 

I think this is a mistake, as it transitively pulls in unstable libs the consumer may not want.

I propose a policy of never depending on unstable AndroidX library versions. 

If this is a required use case, separate unstable artifacts could be published.

**Comment by ychescale9** [comment]
Thanks for calling this out. I agree we should depend on stable versions of AndroidX for version `1.0.0` of the library.

The only reason I've been able to get away with using the alpha version of AndroidX dependencies is that we're technically not API stable either.

In terms of binary compatibility and the effect of depending on non-stable AndroidX libraries:
- AndroidX libraries won't break backward compatibility without a major version bump (e.g. `ViewPager2` which we have a separate artifact for)
- **FlowBinding's** artifacts use `implementation` to depend on the relevant `AndroidX` dependencies. This means consumers of **FlowBinding** need to explicitly bring in the dependency e.g. `androidx.recyclerview` in order to use a binding for that widget. Only the one specified explicitly by the consumer will be packaged into the apk.
- We are tracking binary compatibility of our public APIs using the [binary-compatibility-validator plugin](https://github.com/Kotlin/binary-compatibility-validator).

That said, since the bindings we provide don't use any of those new APIs (maybe except for the material components bindings), there are no reasons to depend on an alpha / beta version of AndroidX.

Will try to address this soon.

**Comment by svenjacobs** [comment]
Even though FlowBinding is not stable yet I would also prefer if the library does not depend on alpha versions of AndroidX libraries. Latest version of FlowBinding broke my build :( Unfortunately I had to downgrade FlowBinding to previous used version.

**Comment by ychescale9** [comment]
I'll prioritize this.

@svenjacobs just out of curiosity what's the error with your broken build? which version were you using previously? I would think FlowBinding should at least remain source compatible after updating these transitive AndroidX dependencies but apparently not? 🤔

**Comment by svenjacobs** [comment]
I upgraded vom 0.10.2 to 0.11.1. I know that 0.10.2 also depends on alpha versions but it seems that with the newest version some clash was introduced with other dependencies of my project. While the build was still working from terminal, Android Studio showed a lot of errors. It could not access some classes from AndroidX anymore. I did not investigate further because I'm quite busy right now ;)

**Comment by ychescale9** [comment]
Hmm that sounds like a tooling / caching issue in the IDE...

Generally it should be safe to depend on alpha version of AndroidX as long as we don't depend on those new APIs introduced in the alphas.

If we were to depend on `MDC 1.2.0-alpha` and release 1.0 today, you should be able to use `MDC 1.1.0` as none of the bindings use **new** callbacks introduced in `1.2.0`. The `Slider` bindings obviously won't work as the `Slider` component is only introduced in `1.2.0`, but if you used `MDC 1.1.0` you won't have access to this component anyway.

But if `flowbinding-material` offered a binding that uses a new callback API introduced in MDC `1.2.0-alpha` and the component is already available in `1.1.0`, we'll have a problem as you'll get a `NoSuchMethodException` at runtime.

So what we really need to make sure is that we aren't using any of the new callback APIs introduced in the latest alpha.

This is all based on the fact that we are using the `implementation` configuration to depend on the AndroidX / MDC dependencies. The advantage of this approach is that consumer is forced to explicitly declare the AndroidX / MDC dependency they want. In the case of `flowbinding-material`, users who depend on `MDC 1.2.0-alpha` will have access to the new `Slider` bindings and the users who are more conservative can still use MDC 1.1.0 and won't suffer from any bugs introduced in `MDC 1.2.0 alpha` even when `flowbinding` depends on that.

The other approach is for FlowBinding to use the `api` configuration and force the consumers to use at least that version of the AndroidX / Material dependency (gradle will [pick the highest version](https://docs.gradle.org/current/userguide/dependency_resolution.html#sub:resolution-strategy) by default). The advantage with this approach is that we don't need to do the API check mentioned earlier and we know consumers won't get any runtime exception. However we really don't want to depend on `alpha` or even `beta` versions of AndroidX in this case as any bugs in these transitive libraries will affect consumers. What this probably means in the long term is that the library would be slower to support new bindings so people who are comfortable using alpha versions of AndroidX or MDC won't get a binding from this library until the AndroidX / MDC library becomes stable. In our case the **Slider** bindings we currently offer would have to be removed if we release 1.0 today.

I must admit that I'm by no means an expert in binary compatibility and making this kind of tradeoffs. I'll see if I can get some help from people who are more experienced with this before deciding what to do for 1.0.

**Comment by svenjacobs** [comment]
Right, I think you should stick with the `implementation` strategy. For a moment I thought you were using `api` which would force alpha versions onto users of your library. Since FlowBinding provides extensions on top of AndroidX libraries, I think it's a good approach to enforce explicitly specifying those AndroidX libraries in a project.

I still wonder why the IDE had problems with the latest version (even after clean cache) and why `./gradlew :app:dependencies` listed alpha versions of AndroidX libraries...

**Comment by ychescale9** [comment]
>I still wonder why the IDE had problems with the latest version (even after clean cache) and why ./gradlew :app:dependencies listed alpha versions of AndroidX libraries...

That's because my assumptions on `api` vs `implementation` were very wrong!

After experimenting a bit with `api` vs `implementation`, I realized using `implementation` and `api` would have very little difference in practice.

Let's stick with the **MDC material** example. If `flowbinding-material` is the **only** dependency in the user's classpath that transitively depends on `com.google.android.material:material`, then yes the **invisibility** of the `material` dependency enforced by `implementation` would effectively "force" the user to explicitly declare the `material` dependency so that they can have access to the extension from `FlowBinding`. However, users might have another library that also transitively depends on `material` using `api` (e.g. `androidx.navigation-ui`). So if users depends on stable `navigation-ui` which depends on stable `material-1.1.0` and `flowbinding-material` that depends on `material-1.2.0-alpha`, the latter will be chosen by default.

So to avoid exposing transitive libraries we really need to use `compileOnly`. But the problem is there's no way to force the user to use at least certain version of the transitive library e.g. user can end up with a really old version of AndroidX / MDC in the classpath without any of the APIs FlowBinding relies on, they'll get runtime exceptions 😂 .

TLDR, yes we should just depend on the latest stable AndroidX / MDC 😄 . `api` vs `implementation` isn't significant.

The AndroidX ones are fine as we don't really depend on any new APIs in the alphas. MDC is a tough call. Should we wait for 1.2.0 stable or remove those Slider bindings?

**Comment by svenjacobs** [comment]
Oh, I didn't know about this behaviour of Gradle, how sneaky! :o

In this case I think you can keep using alpha versions as long as you use `compileOnly`. I think it's the responsibility of the developer to explicitly specify the required transitive dependencies and keep them in sync with FlowBinding. However this should be well documented (which version of FlowBinding depends on what). Testing should cover any runtime exceptions, right ... right?! 😉 

If you want to switch to stable versions then that's fine, too. You can still provide bindings for alpha versions, you just need to publish different artifacts like `flowbinding-material:1.2.0-alpha` (more work on your side 😉 ). Maybe it's a good idea to synchronize FlowBindings artifact versions with the corresponding AndroidX dependency once FB becomes stable?

**Comment by ychescale9** [comment]
I don't really want to rely on documentation to warn users about potential crashes 😅 . All the different versions of AndroidX and their transitive dependencies are already quite brutal for an average user.

I think we should probably just follow how AndroidX handles this. A stable release will just depends on stable release of transitive dependencies. User doesn't need to know that what they've explicitly declared e.g. 1.0.0 has been overridden by a library's transitive dependency 1.1.0, as they are both stable.

I'm not sure how to go about managing a separate non-stable release train 😂 . The AndroidX repo has a lot of custom plugins / tooling / process to support that which I don't have.

I'll look around a bit to see how other libraries handle this.

**Comment by svenjacobs** [comment]
> I think we should probably just follow how AndroidX handles this. A stable release will just depends on stable release of transitive dependencies

👍

**Comment by ychescale9** [comment]
MDC 1.2.0 just hit beta! Let’s wait for it then.
